src/util/container.h

```diff
@@ -303,5 +304,54 @@ class LRUCache
 	std::list<K> m_queue;
 };
 
+template <typename T, size_t N>
+class RingBuffer
+{
+public:
+	RingBuffer() : m_offset(0)
+	{
+		for (size_t i = 0; i < N; ++i)
+			m_data[i] = T(); // buffer is always full
+	}
+	
+	T& at(size_t n)
+	{
+		if (n >= N)
+			throw std::out_of_range("RingBuffer::at");
+		if (n + m_offset < N)
+			return m_data[n + m_offset];
+		else
+			return m_data[n + m_offset - N];
+	}
```

Comment:
push has:
m_data[m_offset] = t;

Which means that at(0) points to the oldest position (the next one to be erased)... Right?